query/src/main/java/org/infinispan/query/dsl/embedded/impl/JPAFilterIndexingServiceProvider.java

```diff
@@ -0,0 +1,324 @@
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryVisited;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.notifications.cachelistener.event.Event;
+import org.infinispan.notifications.cachelistener.event.impl.EventImpl;
+import org.infinispan.notifications.cachelistener.filter.CacheEventConverter;
+import org.infinispan.notifications.cachelistener.filter.CacheEventFilter;
+import org.infinispan.notifications.cachelistener.filter.FilterIndexingServiceProvider;
+import org.infinispan.notifications.cachelistener.filter.IndexedFilter;
+import org.infinispan.objectfilter.FilterCallback;
+import org.infinispan.objectfilter.FilterSubscription;
+import org.infinispan.objectfilter.Matcher;
+import org.infinispan.objectfilter.impl.FilterResultImpl;
+import org.kohsuke.MetaInfServices;
+
+import java.lang.annotation.Annotation;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * @author anistor@redhat.com
+ * @since 7.2
+ */
+@MetaInfServices
+public class JPAFilterIndexingServiceProvider implements FilterIndexingServiceProvider {
+
+   private final ConcurrentMap<Matcher, FilteringListenerInvocation<?, ?>> filteringInvocations = new ConcurrentHashMap<Matcher, FilteringListenerInvocation<?, ?>>(2);
+
+   private CacheNotifierImpl cacheNotifier;
+
+   private ClusteringDependentLogic clusteringDependentLogic;
+
+   @Inject
+   protected void injectDependencies(CacheNotifier cacheNotifier, ClusteringDependentLogic clusteringDependentLogic) {
+      this.cacheNotifier = (CacheNotifierImpl) cacheNotifier;
+      this.clusteringDependentLogic = clusteringDependentLogic;
+   }
+
+   @Override
+   public void start() {
+   }
+
+   @Override
+   public void stop() {
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryActivated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryCreated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryInvalidated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryLoaded.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryModified.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryPassivated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryRemoved.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryVisited.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryEvicted.class).removeAll(filteringInvocations.values());
```

Comment:
I wonder can this use CacheEntriesEvicted instead?  I am hoping to finally remove this event in 8.0 when I redo expiration as well.